### DISCUSSION TITLE
[@types/rails__activestorage] Incorrect type definition for direct upload

### DISCUSSION THREAD

**Discussion by Oleksii14** [DefinitelyTyped/DefinitelyTyped#59071] [Issues with a @types package]
### Steps to reproduce

1. Open `@types/rails__activestorage/index.d.ts`
2. Open `@rails/activestorage/src/direct_upload.js`
3. See that the class constructor is different from what we have in the type definition file.

### Expected behavior
The type definition from step 1 should be correct and implement the class from step 2

### Actual behavior
The type definition from step 1 should be correct and related to the class from step 2. It causes TS errors. And if those errors are fixed according to the incorrect type definition, the DirectUpload class does not work as expected

[a comment by an automated account is not shown]

**Reply by imustafin** [reply]
Hi!

I will take a look later today. What do you mean by "doesn't work as expected"?

Also, which Rails package versions are you using?

As always, you can open a PR and contribute fixes.

**Reply by Oleksii14** [reply]
By "doesn't work as expected" I mean that after I fix the TS errors (that are actually caused by incorrect type definition) I provide the incorrect arguments. And when I run the code like this, the logic is not working

**Reply by imustafin** [reply]
I think I got you.

Indeed, there is an incompatibility of this DT module and Rails 7.0.2 because of https://github.com/rails/rails/commit/aaa64687e81873b26c964fada0b969621ae8cf8f

Can you please check if the previous version of `@types/rails__activestorage` (namely 6.0.1) works? It should work.

The only difference is https://github.com/DefinitelyTyped/DefinitelyTyped/pull/57781. It seems we need to undo the changes and release a new version 7.0.2.

@AlfonsoUceda since you did #57781, do you agree? Have you tried using the current typings with Rails 7.0.2?

**Reply by Oleksii14** [reply]
For now, I just put @ts-ignore and added a valid JS code. I will wait for the next version of the typings. Keep me posted 👍🏻

**Reply by AlfonsoUceda** [reply]
@imustafin, I rolled back to version 6 of this package. The reason is that feature is going to be rewritten so better not to add any new type until they decide what they do.

**Reply by imustafin** [reply]
@AlfonsoUceda in this case I will make a PR tomorrow with the revert to resolve this discussion. It will target rails 7.0.2. After they rewrite the feature we can return to this.

Thanks everyone! Stay tuned for the PR, it would be great if you could review it later.

**Reply by imustafin** [reply]
The PR: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/59098

Take a look please

**Reply by imustafin** [reply]
Thanks for the review!

@Oleksii14 new version should be published soon. Can you please check if it fixes your issues?

**Reply by Oleksii14** [reply]
Yes, fully

**Comment by imustafin** [comment]
For `rails >= 7.0.2`, use `@types/rails__activestorage 7.0.1`
For `rails >= 7.0.0, < 7.0.2`, use `@types/rails__activestorage 7.0.0`
